File: src/opentaiji/providers/anthropic.py
"""
Anthropic Provider
"""

import os
from collections.abc import AsyncGenerator
from typing import Any

from opentaiji.providers.base import LLMProvider, LLMResponse


class AnthropicProvider(LLMProvider):
    """Anthropic Claude Provider"""

    def __init__(self, api_key: str | None = None, model: str = "claude-sonnet-4-20250514", **kwargs):
        super().__init__(api_key=api_key, model=model, **kwargs)
        self.api_key = api_key or os.getenv("ANTHROPIC_API_KEY")
        self.client = None

    def _get_client(self):
        """获取客户端"""
        if self.client is None:
            try:
                from anthropic import AsyncAnthropic

                self.client = AsyncAnthropic(api_key=self.api_key)
            except ImportError as e:
                raise ImportError("anthropic package not installed: pip install anthropic") from e
        return self.client

    async def chat(
        self,
        messages: list[dict],
        tools: list[dict] | None = None,
        temperature: float = 0.7,
        max_tokens: int = 4096,
        stream: bool = False,
        **kwargs,
    ) -> LLMResponse:
        """发送聊天请求"""
        client = self._get_client()

        # 转换消息格式
        formatted_messages: list[dict[str, Any]] = []
        for msg in messages:
            if msg["role"] == "system":
                formatted_messages.insert(0, {"role": "user", "content": f"[System] {msg['content']}"})
            else:
                formatted_messages.append(msg)

        # 构建请求
        request_params = {
            "model": self.model,
            "messages": formatted_messages,
            "temperature": temperature,
            "max_tokens": max_tokens,
        }

        if tools:
            request_params["tools"] = tools

        try:
            response = await client.messages.create(**request_params)

            # 解析响应
            content_blocks = response.content

            text_content = None
            tool_calls = None

            for block in content_blocks:
                if block.type == "text":
                    text_content = block.text
                elif block.type == "tool_use":
                    if tool_calls is None:
                        tool_calls = []
                    tool_calls.append(
                        {
                            "name": block.name,
                            "arguments": block.input,
                            "id": block.id,
                        }
                    )

            return LLMResponse(
                content=text_content,
                tool_calls=tool_calls,
                usage={
                    "input_tokens": response.usage.input_tokens,
                    "output_tokens": response.usage.output_tokens,
                },
                model=self.model,
                raw=response,
            )
        except Exception as e:
            return LLMResponse(
                content=f"Error: {str(e)}",
                raw=response if "response" in locals() else None,
            )

    async def stream_chat(
        self,
        messages: list[dict],
        tools: list[dict] | None = None,
        temperature: float = 0.7,
        max_tokens: int = 4096,
        **kwargs,
    ) -> AsyncGenerator[str, None]:
        """流式聊天"""
        client = self._get_client()

        formatted_messages: list[dict[str, Any]] = []
        for msg in messages:
            if msg["role"] == "system":
                formatted_messages.insert(0, {"role": "user", "content": f"[System] {msg['content']}"})
            else:
                formatted_messages.append(msg)

        request_params: dict[str, Any] = {
            "model": self.model,
            "messages": formatted_messages,
            "temperature": temperature,
            "max_tokens": max_tokens,
            "stream": True,
        }

        if tools:
            request_params["tools"] = tools

        async with client.messages.stream(**request_params) as stream:
            async for event in stream:
                if event.type == "content_block_delta":
                    if hasattr(event.delta, "text"):
                        yield event.delta.text
                    elif hasattr(event.delta, "partial_json"):
                        yield event.delta.partial_json

    def estimate_tokens(self, text: str) -> int:
        """估算 token 数量 (粗略)"""
        return len(text) // 4

File: .repos/harness/packages/core/src/providers/anthropic.ts
/**
 * Anthropic provider - Claude models via the Anthropic Messages API.
 */

import type {
  LLMProvider,
  ChatRequest,
  ChatChunk,
  Message,
} from "./provider.js";

export interface AnthropicProviderConfig {
  apiKey: string;
  baseUrl?: string;
  defaultModel?: string;
}

export class AnthropicProvider implements LLMProvider {
  id = "anthropic";
  name = "Anthropic Claude";
  supportsTools = true;
  supportsStreaming = true;
  supportsImages = true;

  private apiKey: string;
  private baseUrl: string;

  constructor(config: AnthropicProviderConfig) {
    this.apiKey = config.apiKey;
    this.baseUrl = (
      config.baseUrl || "https://api.anthropic.com"
    ).replace(/\/$/, "");
  }

  async *chat(request: ChatRequest): AsyncGenerator<ChatChunk> {
    const body = this.buildRequestBody(request);

    const response = await fetch(`${this.baseUrl}/v1/messages`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "x-api-key": this.apiKey,
        "anthropic-version": "2023-06-01",
      },
      body: JSON.stringify(body),
    });

    if (!response.ok) {
      const errorText = await response.text();
      yield {
        type: "error",
        error: `Anthropic API error ${response.status}: ${errorText}`,
      };
      return;
    }

    if (!response.body) {
      yield { type: "error", error: "No response body" };
      return;
    }

    // Parse SSE stream
    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";
    let currentToolId = "";
    let currentToolName = "";
    let currentToolArgs = "";
    let inputTokens = 0;
    let outputTokens = 0;

    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split("\n");
        buffer = lines.pop() || "";

        for (const line of lines) {
          const trimmed = line.trim();
          if (!trimmed || !trimmed.startsWith("data: ")) continue;

          try {
            const json = JSON.parse(trimmed.slice(6));

            switch (json.type) {
              case "message_start":
                if (json.message?.usage) {
                  inputTokens = json.message.usage.input_tokens || 0;
                }
                break;

              case "content_block_start":
                if (json.content_block?.type === "tool_use") {
                  currentToolId = json.content_block.id || "";
                  currentToolName = json.content_block.name || "";
                  currentToolArgs = "";
                }
                break;

              case "content_block_delta":
                if (json.delta?.type === "text_delta") {
                  yield { type: "text", content: json.delta.text };
                } else if (json.delta?.type === "input_json_delta") {
                  currentToolArgs += json.delta.partial_json || "";
                }
                break;

              case "content_block_stop":
                if (currentToolName) {
                  let args: Record<string, unknown> = {};
                  try {
                    args = JSON.parse(currentToolArgs || "{}");
                  } catch {
                    // Invalid args
                  }
                  yield {
                    type: "tool_call",
                    toolCall: {
                      id: currentToolId,
                      name: currentToolName,
                      args,
                    },
                  };
                  currentToolId = "";
                  currentToolName = "";
                  currentToolArgs = "";
                }
                break;

              case "message_delta":
                if (json.usage) {
                  outputTokens = json.usage.output_tokens || 0;
                }
                break;

              case "message_stop":
                yield {
                  type: "done",
                  usage: { inputTokens, outputTokens },
                };
                break;

              case "error":
                yield {
                  type: "error",
                  error: json.error?.message || "Unknown Anthropic error",
                };
                break;
            }
          } catch {
            // Skip unparseable lines
          }
        }
      }
    } finally {
      reader.releaseLock();
    }
  }

  private buildRequestBody(request: ChatRequest): Record<string, unknown> {
    // Anthropic separates system prompt from messages
    let systemPrompt = "";
    const messages: Array<Record<string, unknown>> = [];

    for (const msg of request.messages) {
      if (msg.role === "system") {
        systemPrompt += (systemPrompt ? "\n" : "") + msg.content;
        continue;
      }

      if (msg.role === "tool") {
        // Anthropic tool results are user messages with tool_result content
        messages.push({
          role: "user",
          content: [
            {
              type: "tool_result",
              tool_use_id: msg.toolCallId,
              content: msg.content,
            },
          ],
        });
        continue;
      }

      if (msg.role === "assistant" && msg.toolCalls?.length) {
        // Assistant message with tool use blocks
        const content: Array<Record<string, unknown>> = [];
        if (msg.content) {
          content.push({ type: "text", text: msg.content });
        }
        for (const tc of msg.toolCalls) {
          content.push({
            type: "tool_use",
            id: tc.id,
            name: tc.name,
            input: tc.args,
          });
        }
        messages.push({ role: "assistant", content });
        continue;
      }

      messages.push({ role: msg.role, content: msg.content });
    }

    const body: Record<string, unknown> = {
      model: request.model,
      messages,
      stream: true,
      max_tokens: request.maxTokens || 4096,
    };

    if (systemPrompt) {
      body.system = systemPrompt;
    }

    if (request.temperature !== undefined) {
      body.temperature = request.temperature;
    }

    if (request.tools?.length) {
      body.tools = request.tools.map((t) => ({
        name: t.name,
        description: t.description,
        input_schema: t.parameters,
      }));
    }

    return body;
  }
}
